nemo_aligner/data/nlp/datasets.py

```diff
@@ -353,6 +353,91 @@ def __getitem__(self, idx):
+       each, which is the response tokens with -100 for the prompt part.
+       
+       WARNING: This class will tokenize the text, but it will raise an exception on model max seq len violations!
+                Meaning it will not truncate tokens to fit to model max seq len, because of special prefix/suffix
+                strings such as <extra_id_1>, it would not know where it is safe to truncate for each model. Therefore,
+                the user must do all truncation logic in their preprocessing step when generating the jsonl
+                used by this class. Put all special truncation logic there specific to your model.
+    """
+
+    def __init__(
+        self, cfg, tokenizer, name, data_prefix, documents, data, seq_length, seed, drop_last=True,
+    ):
+        super().__init__()
+        self.cfg = cfg
+        self.name = name
+        self.data = data
+        self.drop_last = drop_last
+        self.seq_length = seq_length
+        self.tokenizer = tokenizer
+
+        self.reset_position_ids = cfg.data.get("reset_position_ids", False)
+        self.reset_attention_mask = cfg.data.get("reset_attention_mask", False)
+        self.eod_mask_loss = cfg.data.get("eod_mask_loss", False)
+        self.eos_id = tokenizer.eos_id
+
+        np_rng = np.random.default_rng(seed=seed)
+        np_rng.shuffle(self.data)
+
+        self.nograd_length = 32
+
+        # Checks
+        assert np.min(documents) >= 0
+        assert np.max(documents) < len(self.data)
+
+    def __len__(self):
+        return len(self.data)
+
+    def encode(self, text, append_eod=False, add_dummy_prefix=True):
+        if self.cfg.data.get("apply_ftfy", False):
+            import ftfy
+
+            text = ftfy.fix_text(text)
+
+        text_ids = self.tokenizer.text_to_ids(text)
+
+        if len(text_ids) > 0 and append_eod:
+            text_ids.append(self.tokenizer.eos_id)
+
+        return text_ids, len(text_ids)
+
+    def __getitem__(self, idx):
+        """Returns a sample = prompt + response, their respective lengths, and labels.
+        Differently from DPO need to separate the prompt from the response.
+        """
+        payload = self.data[idx]
+        prompt, prompt_len = self.encode(payload["prompt"], append_eod=False)
+        sample, sample_len = self.encode(
+            payload["prompt"] + payload["response"], append_eod=self.cfg.data.get("append_eod", False)
+        )
+        labels = ([-100] * prompt_len) + sample[prompt_len:]
+        # Separate the response from the prompt
+        response = sample[prompt_len:]
+        preference = 1 if payload["preference"] == "chosen" else 0
+
+        assert sample[0:prompt_len] == prompt, "the tokenizer for KTO has merged tokens between prompt and response"
+
+        assert (
+            sample_len <= self.seq_length
+        ), "tokenized text exceeds max seq len! truncate your data in preprocessing prior to KTO training"
```

Comment:
I would recommend you amend this part to do the truncation automatically similar to how we now do it for DPO: https://github.com/NVIDIA/NeMo-Aligner/blob/f1fa2dcaee9b19deee0d2d33b9b6e3251ab7bac7/nemo_aligner/data/nlp/datasets.py#L335

Reply:
Added